src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicCodeCoverageTests.cs

```diff
@@ -0,0 +1,626 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Reflection.PortableExecutable;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Microsoft.CodeAnalysis.Emit;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.DynamicAnalysis.UnitTests
+{
+    public class DynamicCodeCoverageTests : CSharpTestBase
+    {
+        [Fact]
+        public void GotoCoverage()
+        {
+            string source = @"
+using System;
+
+public class Program
+{
+    public static void Main(string[] args)
+    {
+        TestMain();
+    }
+
+    static void TestMain()
+    {
+        Console.WriteLine(""foo"");
+        goto bar;
+        Console.Write(""you won't see me"");
+        bar: Console.WriteLine(""bar"");
+        Fred();
+        return;
+    }
+
+    static void Wilma()
+    {
+        Betty(true);
+        Barney(true);
+        Barney(false);
+        Betty(true);
+    }
+
+    static int Barney(bool b)
+    {
+        if (b)
+            return 10;
+        if (b)
+            return 100;
+        return 20;
+    }
+
+    static int Betty(bool b)
+    {
+        if (b)
+            return 30;
+        if (b)
+            return 100;
+        return 40;
+    }
+
+    static void Fred()
+    {
+        Wilma();
+    }
+}
+
+namespace Microsoft.CodeAnalysis.Runtime
+{
+    public class Instrumentation
+    {
+        private static bool[][] _payloads;
+        private static System.Guid _mvid;
+
+        public static void CreatePayload(System.Guid mvid, int methodToken, ref bool[] payload, int payloadLength)
+        {
+            if (_mvid != mvid)
+            {
+                _payloads = null;
```

Comment:
No, this isn't prototype code for the real implementation--it's just test scaffolding.

CompareExchange is not strictly necessary here. An early iteration of the synthesized instrumentation code did a CompareExchange inline. When I moved that to the helper method, I kept the code in the same form because it matched how I was thinking. If it's bothersome we can minimize the code.